packages/components/src/popover/index.js

```diff
@@ -202,33 +209,55 @@ const Popover = (
 
 	const middleware = [
 		frameOffset || offset
-			? offsetMiddleware( ( { placement: currentPlacement } ) => {
+			? offsetMiddleware( ( middlewareArguments ) => {
 					if ( ! frameOffset ) {
 						return offset;
 					}
-
-					const isTopBottomPlacement =
-						currentPlacement.includes( 'top' ) ||
-						currentPlacement.includes( 'bottom' );
-
+					const {
+						placement: currentPlacement,
+						rects: {
+							reference: referenceRect,
+							floating: floatingRect,
+						},
+					} = middlewareArguments;
 					// The main axis should represent the gap between the
 					// floating element and the reference element. The cross
 					// axis is always perpendicular to the main axis.
-					const mainAxis = isTopBottomPlacement ? 'y' : 'x';
+					const mainAxis = isTopBottomPlacement( currentPlacement )
+						? 'y'
+						: 'x';
 					const crossAxis = mainAxis === 'x' ? 'y' : 'x';
 
 					// When the popover is before the reference, subtract the offset,
 					// of the main axis else add it.
-					const hasBeforePlacement =
-						currentPlacement.includes( 'top' ) ||
-						currentPlacement.includes( 'left' );
-					const mainAxisModifier = hasBeforePlacement ? -1 : 1;
+					const mainAxisModifier = hasBeforePlacement(
+						currentPlacement
+					)
+						? -1
+						: 1;
 					const normalizedOffset = offset ? offset : 0;
 
+					let totalOffset =
+						frameOffset[ mainAxis ] + normalizedOffset;
+					const mainDimension = mainAxis === 'y' ? 'height' : 'width';
+					if (
+						// TODO: does this make sense only for top positioned popovers?
+						currentPlacement.startsWith( 'top' ) &&
+						// If the reference has no height we don't need to do anything.
+						referenceRect[ mainDimension ] &&
+						referenceRect[ mainAxis ] < 0 &&
+						ownerDocument.documentElement.scrollTop <=
```

Comment:
Just flagging that reading the `scrollTop` property could cause reflow / layout trashing, which could affect the page's performance (especially since this middleware is computed as the user scrolls and interacts with the page)